### PR TITLE
Change request-response protocol names to include genesis hash & fork id

### PR DESCRIPTION
**Pull request by dmitry-markin**
## Description

This is a follow-up PR to https://github.com/paritytech/substrate/pull/11938 to rename Polkadot request-response protocols similarly to Substrate changes, originally discussed in https://github.com/paritytech/substrate/issues/7746.

## Open questions

This PR doesn't cover all Polkadot protocols --- changing `validation` & `collation` notification protocols requires a bit of refactoring, so I'm going to create a different PR for them. Please let me know if I missed some other protocols.

### CODE REVIEW

File: node/network/protocol/src/request_response/mod.rs [hunk 1]
```diff
@@ -126,48 +126,53 @@ impl Protocol {
 	///
 	/// Returns a receiver for messages received on this protocol and the requested
 	/// `ProtocolConfig`.
-	pub fn get_config(self) -> (mpsc::Receiver<network::IncomingRequest>, RequestResponseConfig) {
-		let p_name = self.into_protocol_name();
+	pub fn get_config<Hash: AsRef<[u8]>>(
+		self,
+		genesis_hash: &Hash,
+		fork_id: &Option<String>,
```
Comment (bkchr) [inline node/network/protocol/src/request_response/mod.rs:132]:
```suggestion
		fork_id: Option<&str>,
```

Same everywhere else.

**Review by bkchr — COMMENTED** [review]
Some nitpicks, but looks good in general.

File: node/network/bridge/src/network.rs [hunk 2]
```diff
@@ -198,7 +204,10 @@ impl Network for Arc<NetworkService<Block, Hash>> {
 		NetworkService::start_request(
 			&*self,
 			peer_id,
-			protocol.into_protocol_name(),
+			req_protocols
```
[the remark this answers is not in the document]
Comment (dmitry-markin) [reply node/network/bridge/src/network.rs]: Done

File: node/network/protocol/src/request_response/mod.rs (hunk 1 above)
Comment (dmitry-markin) [reply node/network/protocol/src/request_response/mod.rs]: Done

File: node/network/protocol/src/request_response/mod.rs [hunk 3]
```diff
@@ -243,13 +248,13 @@ impl Protocol {
 		}
 	}
 
-	/// Get the protocol name of this protocol, as understood by substrate networking.
-	pub fn into_protocol_name(self) -> Cow<'static, str> {
-		self.get_protocol_name_static().into()
+	/// Fallback protocol names of this protocol, as understood by substrate networking.
+	pub fn get_fallback_names(self) -> Vec<Cow<'static, str>> {
```
Comment (eskimor) [inline node/network/protocol/src/request_response/mod.rs:252]: I am guessing this ensures backwards compatibility?

File: node/service/src/lib.rs [hunk 4]
```diff
@@ -860,17 +857,25 @@ where
 		config.network.extra_sets.extend(peer_sets_info(is_authority));
 	}
 
-	let (pov_req_receiver, cfg) = IncomingRequest::get_config_receiver();
+	let fork_id = config.chain_spec.fork_id().map(ToOwned::to_owned);
+
+	let (pov_req_receiver, cfg) =
+		IncomingRequest::get_config_receiver(&genesis_hash, fork_id.as_deref());
```
Comment (eskimor) [inline node/service/src/lib.rs:863]: Nit: I would actually like it better if we used the new `ReqProtocolNames` as the source of truth. So we only pass in genesis and fork id once into `ReqProtocolNames::new()` and then use that `ReqProtocolNames` everywhere.

**Review by eskimor — APPROVED** [review]
Looks good! @slumber heads up: this will likely cause conflicts with your PR.

File: node/service/src/lib.rs (hunk 4 above)
Comment (bkchr) [reply node/service/src/lib.rs:863]: Very good idea!

And then we could also pass this directly to the overseer below and don't need to pass `fork_id`!

**Review by bkchr — APPROVED** [review]
Besides that last nitpicks mentioned by @eskimor it looks good!

Ty for this :)

**Comment by ordian** [comment]
How do we make sure not to have a conflicting protocol name with substrate?  E.g. if someone later uses `req_chunk` in substrate for something else? Do we expect people to look at both repositories and know where to look? Would it make sense to retain the `/polkadot` prefix after `/}genesis_hash}`?

**Comment by bkchr** [comment]
> How do we make sure not to have a conflicting protocol name with substrate? E.g. if someone later uses `req_chunk` in substrate for something else? Do we expect people to look at both repositories and know where to look? Would it make sense to retain the `/polkadot` prefix after `/{genesis_hash}`?

https://github.com/paritytech/substrate/blob/a1a9b475c354d873f91135ed5fa028ac9ef6a2c4/client/network/src/request_responses.rs#L213

File: node/network/protocol/src/request_response/mod.rs (hunk 3 above)
Comment (dmitry-markin) [reply node/network/protocol/src/request_response/mod.rs:252]: Yes, I think we can at some point obsolete protocol names starting with "/polkadot/..."

File: node/service/src/lib.rs (hunk 4 above)
Comment (dmitry-markin) [reply node/service/src/lib.rs]: Done! About to merge.

**Pull request merged**